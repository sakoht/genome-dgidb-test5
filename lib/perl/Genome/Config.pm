package Genome::Config;

use strict;
use warnings;

our $VERSION = $Genome::VERSION;

BEGIN {
    if (my $config = $ENV{GENOME_CONFIG}) {
        # call the specified configuration module;
        eval "use Genome::Config::$config";
        die $@ if $@;
    }
    else {
        # look for a config module matching all or part of the hostname 
        use Sys::Hostname;
        my $hostname = Sys::Hostname::hostname();
        my @hwords = reverse split('\.',$hostname);
        while (@hwords) {
            my $pkg = 'Genome::Config::' . join("::",@hwords);
            local $SIG{__DIE__};
            local $SIG{__WARN__};
            eval "use $pkg";
            if ($@) {
                pop @hwords;
                next;
            }
            else {
                last;
            }
        }
    }
}
use UR;
use Sys::Hostname;

# This module potentially conflicts to the perl-supplied Config.pm if you've
# set up your @INC or -I options incorrectly.  For example, you used -I /path/to/modules/Genome/
# instead of -I /path/to/modules/.  Many modules use the real Config.pm to get info and
# you'll get wierd failures if it loads this module instead of the right one.
{
    my @caller_info = caller(0);
    if ($caller_info[3] eq '(eval)' and $caller_info[6] eq 'Config.pm') {
        die "package Genome::Config was loaded from a 'use Config' statement, and is not want you wanted.  Are your \@INC and -I options correct?";
    }
}

my $arch_os;
sub arch_os {
    unless ($arch_os) {
        $arch_os = `uname -m`;
        chomp($arch_os);
    }
    return $arch_os;
}

# in dev mode we use dev search, dev wiki, dev memcache, etc, but production database still ;)
my $dev_mode = exists $ENV{GENOME_DEV_MODE} ? $ENV{GENOME_DEV_MODE} : (UR::DBI->no_commit ? 1 : 0);
if ($dev_mode) {
    my $h = hostname;
    warn "***** GENOME_DEV_MODE ($h) *****";
}

sub dev_mode {
    shift;
    if (@_ && !$ENV{GENOME_DEV_MODE}) {
        $dev_mode = shift;
    }

    return $dev_mode;
}

sub base_web_uri {

    if (Genome::Config::dev_mode()) {
        return 'https://aims-dev.gsc.wustl.edu/view';
    } else {
        return 'https://imp.gsc.wustl.edu/view';
    }
}

sub user_email {
    my $self = shift;
    my $user = shift;
    $user ||= $ENV{USER};
    return $user . '@genome.wustl.edu';
}

sub current_user_id {
    return $ENV{USER} . '@genome.wustl.edu';
}

sub admin_notice_users {
    qw/abrummet boberkfe jeldred jlolofie ssmith apipe-run/;
}

sub namespaces {
#    my @ns = (qw/BAP Command EGAP GAP Genome MGAP PAP UR Workflow/);
    my @ns = (qw/Genome UR Workflow/);
    return @ns;
}

# operating directories

sub deploy_path {

    if (Genome::Config::dev_mode()) {
        return '/tmp/gsc/scripts/lib/perl';
    } else {
        return '/gsc/scripts/lib/perl';
    }
}

sub snapshot_paths {

    my @snapshot_paths = (
        '/gsc/scripts/opt/passed-model-tests',
        '/gsc/scripts/opt/passed-unit-tests'
    );

    return @snapshot_paths;
}

sub reference_sequence_directory {
    return join('/', Genome::Config::root_directory(), 'reference_sequences');
}

sub root_directory {
    $ENV{GENOME_MODEL_ROOT} || '/gscmnt/sata420/info/symlinks';
}

sub data_directory {
     $ENV{GENOME_MODEL_DATA} || '/gscmnt/sata835/info/medseq';
}

# links

sub alignment_links_directory {
    return shift->root_directory . '/alignment_links';
}

sub model_comparison_link_directory {
    return shift->root_directory . '/model_comparison_links';
}

# data

sub model_data_directory {
    my $self = shift;
    if (defined($ENV{'GENOME_MODEL_TESTDIR'}) &&
        -e $ENV{'GENOME_MODEL_TESTDIR'}
    ) {
            return $ENV{'GENOME_MODEL_TESTDIR'};
    } else {
            return $self->data_directory .'/model_data';
    }   
}

sub alignment_data_directory {
    return shift->data_directory . '/alignment_data';
}

sub model_comparison_data_directory {
    return shift->data_directory . '/model_comparison_data';
}

# reflection of the different types of models, and their related processing profiles and builds

sub type_names {
    return 
        map { s/\-/ /g; $_ }
        map { Command->_command_name_for_class_word($_) }
        map { s/^Genome\::Model:://; $_ } 
        shift->model_subclass_names;
}

my $use_model_subclasses = 0;
sub _use_model_subclasses {
    # We follow a naming convention which allows us to dynamically list all sub-classes of model.
    # There is some flexibility loss by enforcing the naming convention, but the benefit is reflection.
    # A different config could make a different choice if necessary...
    
    unless ($use_model_subclasses) {
        require Genome::Model;
        my $path = $INC{'Genome/Model.pm'};
        unless ($path) {
            die "failed to find the path for Genome/Model.pm in %INC???";
        }
        $path =~ s/.pm\s*$// or die "no pm on $path?";
        unless (-d $path) {
            die "$path is not a directory?";
        }
        my @possible_subclass_modules = glob("$path/*.pm");
        for my $possible_module (@possible_subclass_modules) {
            my $class = $possible_module;
            $class =~ s/.pm$//;
            $class =~ s/\//\:\:/g;
            $class =~ s/^.*(Genome::Model::[^\:]+)/$1/;
            eval "use $class";
            die "Error using module $class ($possible_module): $@" if $@;
            unless ($class->isa("Genome::Model")) {
                next;
            }
            my $suffix = $class;
            $suffix =~ s/^Genome\::Model:://;
            #$model_subclass_names, $class;
        }
        $use_model_subclasses = 1;
    }
    return 1;
}

1;

=pod

=head1 NAME

Genome::Config - environmental configuration for the genome modeling tools

=head1 DESCRIPTION

The methods in this module are undergoing heavy refactoring and should be ignored until a later release.

=head1 AUTHORS

This software is developed by the analysis and engineering teams at 
The Genome Center at Washington Univiersity in St. Louis, with funding from 
the National Human Genome Research Institute.

=head1 LICENSE

This software is copyright Washington University in St. Louis.  It is released under
the Lesser GNU Public License (LGPL) version 3.  See the associated LICENSE file in
this distribution.

=head1 BUGS

For defects with any software in the genome namespace,
contact genome-dev@genome.wustl.edu.

=cut

