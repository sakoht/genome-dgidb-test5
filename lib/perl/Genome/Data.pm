package Genome::Data;

use strict;
use warnings;

sub create {
    die ("Genome::Data::create must be implemented by the child class");
}
# This class exists just so you can generically refer to sequences, variants
# etc under Genome/Data*. It also makes it possible for DataSets to be composed
# of Data, which seems more appropriate than having DataSets composed of a
# collection of objects of varying types all under Genome/Data. 

1;

