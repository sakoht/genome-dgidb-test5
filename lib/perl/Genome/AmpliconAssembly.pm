package Genome::AmpliconAssembly;

use strict;
use warnings;

use Genome;

use Carp 'confess';
use Cwd 'abs_path';
use Data::Dumper 'Dumper';
require File::Copy;
require Genome::Model::Tools::Consed::Directory;
require Genome::AmpliconAssembly::Amplicon;

my %ATTRIBUTES = (
    directory => {
        is => 'Text',
        is_optional => 1,
        doc => 'Base directory for the amplicon assembly.',
    },
    description => {
        is => 'Text',
        is_optional => 1,
        doc => 'A brief description of the amplicon assembly. Helps identify in messaging. Default is the directory.',
    },
    sequencing_center => {
        is => 'Text',
        is_optional => 1,
        default_value => __PACKAGE__->default_sequencing_center,
        doc => 'Sequencing Center that the amplicons were sequenced. Currently supported centers: '.join(', ', __PACKAGE__->valid_sequencing_centers).' (default: '.__PACKAGE__->default_sequencing_center.').',
    },
    sequencing_platform => {
        is => 'Text',
        is_optional => 1,
        default_value => __PACKAGE__->default_sequencing_platform,
        doc => 'Platform upon whence the amplicons were sequenced. Currently supported platforms: '.join(', ', __PACKAGE__->valid_sequencing_platforms).' (default: '.__PACKAGE__->default_sequencing_platform.').',
    },
    assembly_size => { # amplicon_size
        is => 'Integer',
        is_optional => 1,
        default_value => 1364,
        doc => 'Expected assembly size for an assembled amplicon.',
    },
    subject_name => {
        is => 'Text',
        is_optional => 1,
        doc => 'The subject name of the underlying data. Used primarily in naming files (etc) that represent the entire amplicon assembly, like an assembled fasta of each amplicon.',
    },
    exclude_contaminated_amplicons => {
        is => 'Boolean',
        is_optional => 1,
        default_value => 0,
        doc => 'When getting amplicons, exclude those that have a contaminated read(s). Only for "gsc" generated reads. Default is to include all amplicons.',
    },
    only_use_latest_iteration_of_reads => {
        is => 'Boolean',
        is_optional => 1,
        default_value => 0,
        doc => 'When getting reads for amplicons, only use the most recent iteration for each primer. Only for "gsc" generated reads. Default is to include all reads for each amplicon.',
    },
);

#< UR >#
class Genome::AmpliconAssembly {
    is => 'UR::Object',
    id_by => 'directory',
    has => [
    %ATTRIBUTES,
    _additional_properties => {
        is => 'HASH',
        default_value => {},
        doc => 'Properties added from operations performed on an amplicon assembly.',
    },
    ],
};

sub get { 
    my ($class, %params) = @_;

    # Do not support getting with any other params...
    my $directory = delete $params{directory};
    if ( %params ) { 
        $class->error_message('Can\'t get with additional parameters, only directory.');
        return;
    }
    
    # Validate directory
    unless ( Genome::Sys->validate_existing_directory($directory) ) {
        $class->error_message("Can't validate amplicon assembly directory. See above error.");
        return;
    }

    # UR Cache
    my ($self) = $class->SUPER::get(directory => $directory);
    return $self if $self;
    
    # Properties file
    return $class->_create_from_saved_properties($directory);
}

sub create {
    my ($class, %params) = @_;

    $params{directory} = Cwd::abs_path($params{directory});
    unless ( Genome::Sys->validate_existing_directory($params{directory}) ) {
        $class->error_message("Can't validate amplicon assembly directory. See above error.");
        return;
    }

    # Get UR cache/properties file
    #my ($self) = $class->get(directory => $params{directory});
    #return $self if $self;

    my $self = $class->SUPER::create(%params)
        or return;
    
    unless ( $self->description ) {
        $self->description( $self->directory );
    }

    $self->_validate_properties
        or return;

    $self->create_directory_structure
        or return;
    
    $self->_save_properties
        or return;

    return $self;
}

#< Properties >#
sub _properties_file {
    my ($self, $directory) = @_;

    if ( not $directory and ref $self ) {
        $directory = $self->directory;
    }

    confess "No directory given to get properties file." unless $directory;

    return $directory.'/properties.stor';
}

sub _create_from_saved_properties {
    my ($class, $directory) = @_;

    my $properties_file = $class->_properties_file($directory);
    return unless -e $properties_file;
    
    my $properties = Storable::retrieve($properties_file);
    unless ( $properties ) {
        confess "No properties in amplicon assembly properties file.";
    }

    return $class->SUPER::create(%$properties)
}

sub _validate_properties {
    my $self = shift;

    for my $attribute (qw/ sequencing_platform sequencing_center /) {
        unless ( $self->validate_attribute_value($attribute) ) {
            $self->delete;
            return;
        }
    }

    unless ( $self->sequencing_center eq 'gsc' ) {
        if ( my @attrs = grep { $self->$_ } gsc_only_attributes() ) {
            $self->error_message(
                sprintf(
                    'Indicated unsupported attributes (%s) for sequencing center (%s).',
                    join(', ', @attrs),
                    $self->sequencing_center,
                )
            );
            $self->delete;
            return;
        }
    }

    return 1;
}

sub _save_properties {
    my $self = shift;

    my %properties;
    for my $attr ( $self->attribute_names ) {
        $properties{$attr} = $self->$attr;
    }

    $properties{_additional_properties} = $self->_additional_properties;

    my $properties_file = $self->_properties_file;
    unlink $properties_file if -e $properties_file;
    eval {
        Storable::store(\%properties, $properties_file);
    };
    
    if ( $@ ) {
        confess "Can't save properties to file ($properties_file): $@";
    }
    
    return 1;
}

sub update_additional_properties {
    my ($self, %properties) = @_;

    confess "No properties given to update additional properties." unless %properties;

    for my $property ( keys %properties ) {
        $self->_additional_properties->{$property} = $properties{$property};
    }
    
    return $self->_save_properties;
}

#< Atributes >#
sub attributes {
    return %ATTRIBUTES;
}

sub attribute_names {
    return keys %ATTRIBUTES;
}

sub attributes_without_default_values {
    my $class = shift;

    my %attributes;
    for my $name ( attribute_names() ) {
        $attributes{$name} = $class->get_attribute_for_name($name);
        delete $attributes{$name}->{default_value};
    }

    return %attributes;
}

sub get_attribute_for_name {
    my ($class, $name) = @_;

    confess "No attribute name given to get attribute" unless $name;
    confess "No attribute found for name ($name)" unless exists $ATTRIBUTES{$name};

    return $ATTRIBUTES{$name};
}

sub helpful_methods {
    return (qw/ 
        chromat_dir phd_dir edit_dir
        consed_directory create_directory_structure
        get_amplicons 
        amplicon_fasta_types amplicon_bioseq_method_for_type
        fasta_file_for_type qual_file_for_type
        assembly_fasta reads_fasta processed_fasta 
        /);
}

sub validate_attribute_value {
    my ($self, $attribute) = @_;

    my $value = $self->$attribute;
    unless ( defined $value ) { # all attrs have defaults, make sure it is set
        $self->error_message("No value for $attribute given");
        return;
    }
    
    my $valid_values_method = 'valid_'.$attribute.'s';
    unless ( grep { $value eq $_ } $self->$valid_values_method ) {
        $self->error_message(
            sprintf(
                'Invalid %s %s. Valid %ss: %s',
                $attribute,
                $value,
                $attribute,
                join(', ',valid_sequencing_centers()),
            )
        );
        return;
    }

    return 1;
}

# Sequencing Centers
sub valid_sequencing_centers {
    return (qw/ gsc broad /);
}

sub default_sequencing_center {
    return (valid_sequencing_centers)[0];
}

# Sequencing Platforms
sub valid_sequencing_platforms {
    return (qw/ sanger /);
}

sub default_sequencing_platform {
    return (valid_sequencing_platforms)[0];
}

# GSC Only Attributes
sub gsc_only_attributes {
    return (qw/ only_use_latest_iteration_of_reads exclude_contaminated_amplicons /);
}

#< DEPRACATED FIXME #>
sub assembler { return 'phredphrap' };

#< DIRS >#
sub consed_directory {
    my $self = shift;

    unless ( $self->{_consed_directory} ) {
        $self->{_consed_directory} = Genome::Model::Tools::Consed::Directory->create(directory => $self->directory);
    }

    return $self->{_consed_directory};
}

sub create_directory_structure {
    my $self = shift;

    $self->consed_directory->create_extended_directory_structure
        or return;

    return 1;
}

sub edit_dir {
    return $_[0]->consed_directory->edit_dir;
}
    
sub phd_dir {
    return $_[0]->consed_directory->phd_dir;
}
    
sub chromat_dir {
    return $_[0]->consed_directory->chromat_dir;
}

sub fasta_dir {
    return $_[0]->consed_directory->fasta_dir;
}

#< FASTA >#
my %_fastas_and_amplicon_bioseq_methods = (
    reads => 'get_bioseqs_for_raw_reads',
    processed => 'get_bioseqs_for_processed_reads',
    assembly => 'get_assembly_bioseq',
    oriented => 'get_oriented_bioseq',
);

sub amplicon_fasta_types {
    return keys %_fastas_and_amplicon_bioseq_methods;
}

sub amplicon_bioseq_method_for_type {
    return $_fastas_and_amplicon_bioseq_methods{$_[1]};
}

sub fasta_file_for_type {
    my ($self, $type) = @_;

    return sprintf(
        '%s/%s%s.fasta',
        $self->fasta_dir,
        ( defined $self->subject_name ? $self->subject_name.'.' : '' ),
        $type,
    );
}

sub qual_file_for_type {
    return $_[0]->fasta_file_for_type($_[1]).'.qual';
}

#< Amplicons >#
sub get_amplicons {
    my $self = shift;

    my $method = sprintf(
        '_get_amplicons_and_read_names_for_%s_%s', 
        $self->sequencing_center,
        $self->sequencing_platform,
    );

    my $iterator = $self->$method;
    
    my $amplicons = $self->$method;
    unless ( $amplicons and %$amplicons ) {
        $self->error_message(
            sprintf('No amplicons found in chromat_dir of directory (%s)', $self->directory) 
        );
        return;
    }

    # Only use most recent read
    if ( $self->only_use_latest_iteration_of_reads ) {
        $self->_remove_old_read_iterations_from_amplicons($amplicons)
            or return;
    }
    
    if ( $self->exclude_contaminated_amplicons ) {
        $self->_remove_contaminated_amplicons($amplicons)
            or return;
    }
    
    my @amplicons;
    my $edit_dir = $self->edit_dir;
    for my $name ( sort { $a cmp $b } keys %$amplicons ) {
        push @amplicons, Genome::AmpliconAssembly::Amplicon->create(
            name => $name,
            reads => $amplicons->{$name},
            directory => $edit_dir,
        );
    }

    return \@amplicons;
}

sub _get_amplicons_and_read_names_for_gsc_sanger {
    my $self = shift;

    my $read_iterator = $self->_get_read_name_iterator
        or return;

    my %amplicons;
    while ( my $read_name = $read_iterator->() ) {
        my $amplicon_name = $self->_get_amplicon_name_for_gsc_sanger_read_name($read_name)
            or next; # Assume conversion logic is good, and returns undef cuz it is not a read
        push @{$amplicons{$amplicon_name}}, $read_name;
    }

    return \%amplicons;
}

sub _get_amplicons_and_read_names_for_broad_sanger {
    my $self = shift;

    my $read_iterator = $self->_get_read_name_iterator
        or return;

    my %amplicons;
    while ( my $read_name = $read_iterator->() ) {
        my $amplicon = $self->_get_amplicon_name_for_broad_sanger_read_name($read_name)
            or return; # Assume conversion logic is good, and returns undef cuz it is not a read
        push @{$amplicons{$amplicon}}, $read_name;
    }

    return  \%amplicons;
}

sub _get_read_name_iterator {
    my $self = shift;

    my $dh = Genome::Sys->open_directory( $self->chromat_dir );
    unless ( $dh ) {
        $self->error_message("Can't open chromat dir to get reads. See above error.");
        return;
    }

    return sub{
        while ( my $name = $dh->read ) {
            next if $name =~ m#^\.#;
            $name =~ s#\.gz##;
            return $name;
        }
        $dh->close;
        return;
    }
}

sub _remove_old_read_iterations_from_amplicons {
    my ($self, $amplicons) = @_;

    for my $amplicon_name ( keys %$amplicons ) {
        my %reads;
        for my $read_name ( @{$amplicons->{$amplicon_name}} ) {
            my $read = $self->_get_gsc_sequence_read($read_name);
            confess "Can't get read for name ($read_name). This is required to remove old read iterations from amplicons" unless $read;
            
            my $read_id = $amplicon_name.$read->primer_code;
            if ( exists $reads{$read_id} ) {
                my $date_compare = UR::Time->compare_dates(
                    '00:00:00 '.$read->run_date,
                    '00:00:00 '.$reads{$read_id}->run_date,
                ); #returns -1, 0, or 1
                #print "RUN DATE $read_name => ".$read->run_date."($date_compare)\n";
                $reads{$read_id} = $read if $date_compare eq 1;
            }
            else {
                $reads{$read_id} = $read;
            }
        }
        $amplicons->{$amplicon_name} = [
        sort { $a cmp $b } 
        map { $_->trace_name }
        values %reads 
        ];
    }

    return 1;
}

sub _remove_contaminated_amplicons {
    my ($self, $amplicons) = @_;

    AMPLICON: for my $amplicon_name ( keys %$amplicons ) {
        for my $read_name ( @{$amplicons->{$amplicon_name}} ) {
            my $read = $self->_get_gsc_sequence_read($read_name);
            confess "Can't get read for name ($read_name). This is required to remove contaminated amplicons." unless $read;
            my $screen_reads_stat = $read->get_screen_read_stat_hmp;
            if ( $screen_reads_stat and $screen_reads_stat->is_contaminated ) {
                delete $amplicons->{$amplicon_name};
                next AMPLICON;
            }
        }
    }

    return 1;
}

sub _get_gsc_sequence_read { # in sub to overload on test
    return GSC::Sequence::Read->get(trace_name => $_[1]);
}

#< Amplicon Reads >#
sub get_method_for_get_amplicon_name_for_read_name {
    my $self = shift;

    return sprintf(
        '_get_amplicon_name_for_%s_%s_read_name', 
        $self->sequencing_center,
        $self->sequencing_platform,
    );
}

sub _get_amplicon_name_for_gsc_sanger_read_name {
    my ($self, $read_name) = @_;

    $read_name =~ /^(.+)\.[bg]\d+$/
        or return;

    return $1;
}

sub _get_amplicon_name_for_broad_sanger_read_name {
    my ($self, $read_name) = @_;

    $read_name =~ s#\.T\d+$##;
    $read_name =~ s#[FR](\w\d\d?)$#\_$1#; # or next;

    return $read_name;
}

sub get_all_amplicons_reads_for_read_name {
    my ($self, $read_name) = @_;

    unless ( $read_name ) {
        $self->error_message("No read name given to get all reads for amplicon.");
        return;
    }

    my $amp_method = $self->get_method_for_get_amplicon_name_for_read_name;
    my $amplicon_name = $self->$amp_method($read_name);
    unless ( $amplicon_name ) {
        $self->error_message("Can't get amplicon name for read name ($read_name)");
        return;
    }

    my $reads_method = $self->get_method_for_get_all_amplicons_reads_for_read_name;
    my @read_names = $self->$reads_method($amplicon_name);
    unless ( @read_names ) {
        $self->error_message("No reads found for amplicon name ($amplicon_name)");
        return;
    }


    return @read_names;
}

sub get_method_for_get_all_amplicons_reads_for_read_name {
    sprintf(
        '_get_all_reads_for_%s_%s_amplicon',
        $_[0]->sequencing_center,
        $_[0]->sequencing_platform,
    );
}

sub _get_all_reads_for_gsc_sanger_amplicon {
    my ($self, $amplicon_name) = @_;
    
    my $chromat_dir = $self->chromat_dir;
    my @read_names;
    for my $read_name ( glob("$chromat_dir/$amplicon_name.*") ) {
        $read_name =~ s#$chromat_dir/##;
        $read_name =~ s#\.gz##;
        push @read_names, $read_name;
    }

    return @read_names;
}

sub _get_all_reads_for_broad_sanger_amplicon {
    die "Not implemented\n";
}

#< Contamination Screening >#
sub contamination_dir {
    return $_[0]->directory.'/contamination';
}

sub contamination_reads_dir {
    return $_[0]->contamination_dir.'/reads';
}
    
sub amplicon_fasta_file_for_contamination_screening {
    return $_[0]->contamination_dir.'/amplicon_reads.fasta';
}

sub create_contamination_dir_structure {
    my $self = shift;

    my $contamination_dir = $self->contamination_dir;
    unless ( -d $contamination_dir ) {
        return unless Genome::Sys->create_directory($contamination_dir);
    }

    my $reads_dir = $self->contamination_reads_dir;
    unless ( -d $reads_dir ) {
        return unless Genome::Sys->create_directory($reads_dir);
    }

    return 1;
}

sub create_amplicon_fasta_file_for_contamination_screening {
    my $self = shift;

    my $amplicons = $self->get_amplicons
        or return;

    $self->create_contamination_dir_structure
        or return;
    
    my $fasta_file = $self->amplicon_fasta_file_for_contamination_screening;
    unlink $fasta_file if -e $fasta_file;
    my $fasta_writer = Bio::SeqIO->new(
        '-file' => '>'.$fasta_file,
        '-fomat' => 'fasta',
    );
    for my $amplicon ( @$amplicons ) {
        for my $bioseq ( $amplicon->get_bioseqs_for_processed_reads ) {
            next unless $bioseq->length >= 11;
            $bioseq->seq( uc $bioseq->seq );
            $fasta_writer->write_seq($bioseq);
        }
    }

    return $fasta_file;
}

sub remove_contaminated_amplicons_by_reads_in_file {
    my ($self, $file) = @_;

    $self->create_contamination_dir_structure
        or return;

    my $fh = Genome::Sys->open_file_for_reading($file)
        or return;

    my %amplicons_seen;
    my $chromat_dir = $self->chromat_dir;
    my $contamination_reads_dir = $self->contamination_reads_dir;
    my $amp_method = $self->get_method_for_get_amplicon_name_for_read_name;
    my $reads_method = $self->get_method_for_get_all_amplicons_reads_for_read_name;

    while ( my $read_name = $fh->getline ) {
        chomp $read_name;
        my $amplicon_name = $self->$amp_method($read_name);
        unless ( $amplicon_name ) {
            $self->error_message("Can't get amplicon name for read name ($read_name)");
            return;
        }
        my @read_names = $self->$reads_method($amplicon_name);
        unless ( @read_names ) {
            $self->error_message("Can't get reads for amplicon name ($amplicon_name)");
            return;
        }
        # Move chromats - this effectively removes the amplicon
        # TODO create the amplicon interface here, to support sanger and 454
        #  maybe move this logic to amplicon?
        for my $read_name ( @read_names ) {
            my $from = "$chromat_dir/$read_name.gz"; 
            my $to = "$contamination_reads_dir/$read_name.gz";
            unless ( -e $from ) { # ok, i guess
                $self->error_message("Can't find trace $read_name");
                next;
            }
            unless ( File::Copy::move($from, $to) ) { # not ok
                $self->error_message("Can't move $from to $to\: $!");
                return;
            }
        }
        # TODO move phd_dir edit_dir files?
    }

    return 1;
}

1;

=pod

=head1 Name

Genome::AmpliconAssembly

=head1 Synopsis

=head1 Usage

=head1 Methods

=head2 valid_sequencing_centers

=over

=item I<Synopsis>

=item I<Arguments>

=item I<Returns>

=back

=head2 default_sequencing_center

=over

=item I<Synopsis>

=item I<Arguments>

=item I<Returns>

=back

=head2 consed_directory

=over

=item I<Synopsis>

=item I<Arguments>

=item I<Returns>

=back

=head2 create_directory_structure

=over

=item I<Synopsis>

=item I<Arguments>

=item I<Returns>

=back

=head2 edit_dir

=over

=item I<Synopsis>

=item I<Arguments>

=item I<Returns>

=back

=head2 phd_dir

=over

=item I<Synopsis>

=item I<Arguments>

=item I<Returns>

=back

=head2 chromat_dir

=over

=item I<Synopsis>

=item I<Arguments>

=item I<Returns>

=back

=head2 fasta_dir

=over

=item I<Synopsis>

=item I<Arguments>

=item I<Returns>

=back

=head2 amplicon_fasta_types

=over

=item I<Synopsis>

=item I<Arguments>

=item I<Returns>

=back

=head2 amplicon_bioseq_method_for_type

=over

=item I<Synopsis>

=item I<Arguments>

=item I<Returns>

=back

=head2 fasta_file_for_type

=over

=item I<Synopsis>

=item I<Arguments>

=item I<Returns>

=back

=head2 qual_file_for_type

=over

=item I<Synopsis>

=item I<Arguments>

=item I<Returns>

=back

=head2 get_amplicons

=over

=item I<Synopsis>

=item I<Arguments>

=item I<Returns>

=back

=head2 contamination_dir

=over

=item I<Synopsis>

=item I<Arguments>

=item I<Returns>

=back

=head2 contamination_reads_dir

=over

=item I<Synopsis>

=item I<Arguments>

=item I<Returns>

=back

=head2 amplicon_fasta_file_for_contamination_screening

=over

=item I<Synopsis>

=item I<Arguments>

=item I<Returns>

=back

=head2 create_contamination_dir

=over

=item I<Synopsis>

=item I<Arguments>

=item I<Returns>

=back

=head2 create_amplicon_fasta_files_for_contamination_screening

=over

=item I<Synopsis>

=item I<Arguments>

=item I<Returns>

=back

=head2 read_is_contaminated

=over

=item I<Synopsis>

=item I<Arguments>

=item I<Returns>

=back

=head1 See Also

=head1 Disclaimer

Copyright (C) 2005 - 2009 Genome Center at Washington University in St. Louis

This module is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY or the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.

=head1 Author(s)

B<Eddie Belter> I<ebelter@genome.wustl.edu>

=cut

#$HeadURL$
#$Id$
