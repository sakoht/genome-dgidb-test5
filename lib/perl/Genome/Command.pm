package Genome::Command;

use strict;
use warnings;

use Genome;

class Genome::Command {
    is => 'Command::Tree',
};
         
# This map allows the top-level genome commands to be whatever
# we wish, instead of having to match the directory structure.
my %command_map = (
    'disk' => 'Genome::Disk::Command',
    'feature-list' => 'Genome::FeatureList::Command',
    'gene-name' => 'Genome::GeneName::Command',
    'individual' => 'Genome::Individual::Command',
    'instrument-data' => 'Genome::InstrumentData::Command',
    'library' => 'Genome::Library::Command',
    'model' => 'Genome::Model::Command',
    'model-group' => 'Genome::ModelGroup::Command',
    'notable' => 'Genome::Notable::Command',
    'population-group' => 'Genome::PopulationGroup::Command',
    'processing-profile' => 'Genome::ProcessingProfile::Command',
    'project' => 'Genome::Project::Command',
    'project-part' => 'Genome::ProjectPart::Command',
    'report' => 'Genome::Report::Command',
    'sample' => 'Genome::Sample::Command',
    'subject' => 'Genome::Subject::Command',
    'sys' => 'Genome::Sys::Command',
    'task' => 'Genome::Task::Command',
    'taxon' => 'Genome::Taxon::Command',
    'tools' => 'Genome::Model::Tools',
    'db' => 'Genome::Db',
);

for my $class (values %command_map) {
    eval "use $class";
    die $@ if $@;
}

$Genome::Command::SUB_COMMAND_MAPPING = \%command_map;

1;
