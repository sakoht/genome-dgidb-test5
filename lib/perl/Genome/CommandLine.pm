package Genome::CommandLine;

use strict;
use warnings;

use Genome;

class Genome::CommandLine {
    is => 'Command::Tree',
};

sub command_tree_source_classes {
    return ('Genome::Model::Tools','Genome::Command');
}

sub _categorize_sub_commands {
    my $class = shift;

    my @sub_command_classes = $class->sorted_sub_command_classes;
    my %categories;
    my @order;
    for my $sub_command_class (@sub_command_classes) {
        next if $sub_command_class->_is_hidden_in_docs();
        #next unless $sub_command_class =~ /Genome::B/;

        my $category;
        if ($sub_command_class =~ /Genome::Model::Tools/) {
            $category = 'tools';
        }
        else {
            $category = 'system';
        }

        unless (exists $categories{$category}) {
            if ($category) {
                push(@order, $category)
            } else {
                unshift(@order, '');
            }
            $categories{$category} = [];
        }
        push(@{$categories{$category}}, $sub_command_class);
    }

    return (\@order, \%categories);
}

1;
