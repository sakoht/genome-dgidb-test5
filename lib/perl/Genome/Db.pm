use strict;
use warnings;
use Genome;

class Genome::Db {
    is => 'Command::Tree',
    doc => 'external database interfaces',
};

1;
