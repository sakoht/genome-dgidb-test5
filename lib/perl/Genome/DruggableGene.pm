package Genome::DruggableGene;

use strict;
use warnings;

use Genome;

class Genome::DruggableGene {
    is => 'UR::Singleton',
};

1;
