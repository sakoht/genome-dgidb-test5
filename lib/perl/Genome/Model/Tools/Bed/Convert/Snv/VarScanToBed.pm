package Genome::Model::Tools::Bed::Convert::Snv::VarScanToBed;

use strict;
use warnings;

use Genome;

class Genome::Model::Tools::Bed::Convert::Snv::VarScanToBed {
    is => ['Genome::Model::Tools::Bed::Convert::Snv'],
};

sub help_brief {
    "Tools to convert var-scan SNV format to BED.",
}

sub help_synopsis {
    my $self = shift;
    return <<"EOS"
  gmt bed convert snv var-scan-to-bed --source snps_all_sequences --output snps_all_sequences.bed
EOS
}

sub help_detail {                           
    return <<EOS
    This is a small tool to take SNV calls in var-scan format and convert them to a common BED format (using the first four columns).
EOS
}

sub process_source {
    my $self = shift;
    
    my $input_fh = $self->_input_fh;
    
    while(my $line = <$input_fh>) {
        my ($chromosome, $position, $reference, $consensus, @extra) = split("\t", $line);
        
        no warnings qw(numeric);
        next unless $position eq int($position); #Skip header line(s)
        use warnings qw(numeric);
        
        #position => 1-based position of the SNV
        #BED uses 0-based position of and after the event
        $self->write_bed_line($chromosome, $position - 1, $position, $reference, $consensus);
    }
    
    return 1;
}

1;
