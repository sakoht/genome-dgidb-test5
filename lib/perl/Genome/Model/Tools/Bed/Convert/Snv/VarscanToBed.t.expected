1	137	138	T/Y	30
1	143	144	T/Y	31
1	179	180	T/Y	31
1	230	231	C/M	33
1	233	234	C/Y	33
1	247	248	A/W	34
1	249	250	A/M	31
1	254	255	A/W	33
1	256	257	A/M	33
1	290	291	C/Y	27
1	314	315	C/Y	28
1	320	321	C/Y	29
1	326	327	T/Y	28
1	327	328	A/W	27
1	389	390	C/M	32
1	392	393	C/Y	34
1	396	397	C/M	32
1	439	440	C/M	34
1	442	443	C/Y	34
1	469	470	G/R	32
1	471	472	G/S	32
1	472	473	G/S	33
1	481	482	G/S	32
1	518	519	G/S	35
1	582	583	G/R	33
1	602	603	G/R	34
1	4769	4770	A/R	35
1	4792	4793	A/G	34
1	4980	4981	A/R	35
1	5052	5053	G/R	33
1	5073	5074	T/G	0
